File: dev_browser_mcp/paths.py
import os
import sys
import time
from pathlib import Path
from typing import Optional


def now_ms() -> int:
    return int(time.time() * 1000)


def platform_cache_dir() -> Path:
    home = Path.home()

    xdg_cache_home = os.environ.get("XDG_CACHE_HOME")
    if xdg_cache_home:
        return Path(xdg_cache_home)

    if sys.platform == "darwin":
        return home / "Library" / "Caches"

    return home / ".cache"


def platform_state_dir() -> Path:
    home = Path.home()

    xdg_state_home = os.environ.get("XDG_STATE_HOME")
    if xdg_state_home:
        return Path(xdg_state_home)

    if sys.platform == "darwin":
        return home / "Library" / "Application Support"

    return home / ".local" / "state"


def safe_artifact_path(artifact_dir: Path, path_arg: Optional[str], default_name: str) -> Path:
    allow_unsafe = os.environ.get("DEV_BROWSER_ALLOW_UNSAFE_PATHS", "").lower() in {"1", "true", "yes"}

    if not path_arg:
        artifact_dir.mkdir(parents=True, exist_ok=True)
        return artifact_dir / default_name

    expanded = Path(os.path.expandvars(os.path.expanduser(path_arg)))
    resolved = expanded if expanded.is_absolute() else (artifact_dir / expanded)
    resolved = resolved.resolve()

    if allow_unsafe:
        resolved.parent.mkdir(parents=True, exist_ok=True)
        return resolved

    artifact_dir_resolved = artifact_dir.resolve()
    if artifact_dir_resolved not in resolved.parents and resolved != artifact_dir_resolved:
        raise ValueError(
            f"Refusing to write outside artifact dir: {resolved} (allowed under {artifact_dir_resolved}). "
            "Set DEV_BROWSER_ALLOW_UNSAFE_PATHS=1 to override."
        )

    resolved.parent.mkdir(parents=True, exist_ok=True)
    return resolved

File: internal/devbrowser/paths.go
package devbrowser

import (
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"
)

const (
	cacheSubdir = "dev-browser-go"
)

func NowMS() int64 {
	return time.Now().UnixNano() / int64(time.Millisecond)
}

func PlatformCacheDir() string {
	if dir := os.Getenv("XDG_CACHE_HOME"); strings.TrimSpace(dir) != "" {
		return dir
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	if runtime.GOOS == "darwin" {
		return filepath.Join(home, "Library", "Caches")
	}
	return filepath.Join(home, ".cache")
}

func PlatformStateDir() string {
	if dir := os.Getenv("XDG_STATE_HOME"); strings.TrimSpace(dir) != "" {
		return dir
	}
	home, err := os.UserHomeDir()
	if err != nil {
		return ""
	}
	if runtime.GOOS == "darwin" {
		return filepath.Join(home, "Library", "Application Support")
	}
	return filepath.Join(home, ".local", "state")
}

func ArtifactDir(profile string) string {
	return filepath.Join(PlatformCacheDir(), cacheSubdir, profile, "artifacts")
}

func StateDir(profile string) string {
	return filepath.Join(PlatformStateDir(), cacheSubdir, profile)
}

func StateFile(profile string) string {
	return filepath.Join(StateDir(profile), "daemon.json")
}

func SafeArtifactPath(artifactDir, pathArg, defaultName string) (string, error) {
	allowUnsafe := envTruthy("DEV_BROWSER_ALLOW_UNSAFE_PATHS")

	if strings.TrimSpace(pathArg) == "" {
		if err := os.MkdirAll(artifactDir, 0o755); err != nil {
			return "", err
		}
		return filepath.Join(artifactDir, defaultName), nil
	}

	expanded := os.ExpandEnv(pathArg)
	if strings.HasPrefix(expanded, "~") {
		if home, err := os.UserHomeDir(); err == nil {
			expanded = filepath.Join(home, strings.TrimPrefix(expanded, "~"))
		}
	}

	resolved := expanded
	if !filepath.IsAbs(resolved) {
		resolved = filepath.Join(artifactDir, resolved)
	}
	resolved, err := filepath.Abs(resolved)
	if err != nil {
		return "", err
	}

	if allowUnsafe {
		if err := os.MkdirAll(filepath.Dir(resolved), 0o755); err != nil {
			return "", err
		}
		return resolved, nil
	}

	artifactResolved, err := filepath.Abs(artifactDir)
	if err != nil {
		return "", err
	}
	allowed := artifactResolved == resolved || strings.HasPrefix(resolved, artifactResolved+string(os.PathSeparator))
	if !allowed {
		return "", fmt.Errorf("Refusing to write outside artifact dir: %s (allowed under %s). Set DEV_BROWSER_ALLOW_UNSAFE_PATHS=1 to override.", resolved, artifactResolved)
	}

	if err := os.MkdirAll(filepath.Dir(resolved), 0o755); err != nil {
		return "", err
	}
	return resolved, nil
}

func envTruthy(name string) bool {
	v := strings.ToLower(strings.TrimSpace(os.Getenv(name)))
	return v == "1" || v == "true" || v == "yes" || v == "on"
}

func clampNonNegativeInt(value int) (int, error) {
	if value < 0 {
		return 0, errors.New("value must be non-negative")
	}
	return value, nil
}
